File: pdf_fill_studio/editor/coords.js
// Editor pixels -> PDF points (top-left origin). `scale` is pixels per point.
export function pxToPoints(px, scale) {
  return px / scale;
}

// Convert an absolutely-positioned box (CSS px, top-left) to PDF points.
export function boxPxToPoints(boxPx, scale) {
  return {
    x: pxToPoints(boxPx.left, scale),
    y: pxToPoints(boxPx.top, scale),
    w: pxToPoints(boxPx.width, scale),
    h: pxToPoints(boxPx.height, scale),
  };
}

File: pdf_fill_studio/coords.py
"""Coordinate conversions. job.json uses PDF points, TOP-LEFT origin (y grows downward)."""

def topleft_box_to_baseline(x, y, w, h, page_height, font_size):
    """Convert a top-left-origin box to a reportlab text anchor (bottom-left origin).

    Returns (x, baseline_y) where baseline_y is where reportlab should draw the text
    so it sits near the bottom of the box with a small descent allowance.
    """
    baseline_y = page_height - y - h + 0.2 * font_size
    return x, baseline_y

def px_to_points(value_px, scale):
    """Editor pixels -> PDF points. `scale` is pixels per point (the pdf.js viewport scale)."""
    return value_px / scale
